lib/monitor.toit

```diff
@@ -35,34 +35,44 @@ A latch that allows one task to wait until a value (object) has been provided
 This class must not be extended.
 */
 monitor Latch:
-  has_value_ / bool := false
+  static STATE_UNSET_         ::= 0
+  static STATE_HAS_VALUE_     ::= 1
+  static STATE_HAS_EXCEPTION_ ::= 2
+
+  state_ / int := STATE_UNSET_
   value_ := null
 
   /**
   Receives the value.
 
-  This method blocks until the value is available.
+  This method blocks until the value is available. If the value
+    is an exception, it is thrown.
   May be called multiple times.
   */
   get -> any:
-    await: has_value_
-    return value_
+    await: state_ != STATE_UNSET_
+    value := value_
+    if state_ == STATE_HAS_EXCEPTION_:
+      if value is not Exception_: throw value
+      rethrow value.value value.trace
+    return value
 
   /**
   Sets the $value of the latch.
 
   Calling this method unblocks any task that is blocked in the $get method of
     the same instance, sending the $value to it.
   Future calls to $get return immediately and use this $value.
-  Must be called at most once for each instance of the monitor.
```

Comment:
We could also decide to enforce this. It might be a nice property in terms of being able to reason about values returned by get.